### PR TITLE
Create cgroups for kube and runtime if configured

### LINKED ISSUE #11607: Enforcing kube-reserved fails after migrating to containerd runtime
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
Tried 1.19.3 and 1.20.1

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
1.19.10 and 1.20.6

**3. What cloud provider are you using?**
AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
Changed spec to add `containerRuntime: containerd`

For the 1.19 setup, also tried specifying `cgroupDriver: systemd` in the kubelet spec

**5. What happened after the commands executed?**
Node fails to start, kubelet crashes with:
```
May 26 13:57:27 ip-172-16-3-109 kubelet[6848]: I0526 13:57:27.795825    6848 node_container_manager_linux.go:121] Enforcing kube reserved on cgroup "/kube-reserved" with limits: map[cpu:{i:{value:750 scale:-3} d:{Dec:<nil>} s:750m Format:DecimalSI} ephemeral-storage:{i:{value:1073741824 scale:0} d:{Dec:<nil>} s:1Gi Format:BinarySI} memory:{i:{value:2147483648 scale:0} d:{Dec:<nil>} s:2Gi Format:BinarySI}]

May 26 13:57:27 ip-172-16-3-109 kubelet[6848]: F0526 13:57:27.795912    6848 kubelet.go:1321] Failed to start ContainerManager Failed to enforce Kube Reserved Cgroup Limits on "/kube-reserved": ["kube-reserved"] cgroup does not exist
```

**6. What did you expect to happen?**

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2019-02-01T15:20:02Z"
  generation: 9
  name: k8s.ops.us-west-2.aws.domain
spec:
  additionalPolicies:
    master: |
      [
        {
          "Effect": "Allow",
          "Action": [
            "autoscaling:DescribeAutoScalingGroups",
            "autoscaling:DescribeAutoScalingInstances",
            "autoscaling:DescribeLaunchConfigurations",
            "autoscaling:DescribeTags",
            "autoscaling:SetDesiredCapacity",
            "autoscaling:TerminateInstanceInAutoScalingGroup"
          ],
          "Resource": ["*"]
        }
      ]
    node: "[\n  { \n    \"Effect\": \"Allow\",\n    \"Action\": [\n      \"ec2:Describe*\",\n
      \     \"ecr:GetAuthorizationToken\",\n      \"ecr:BatchCheckLayerAvailability\",\n
      \     \"ecr:GetDownloadUrlForLayer\",\n      \"ecr:GetRepositoryPolicy\",\n
      \     \"ecr:DescribeRepositories\",\n      \"ecr:ListImages\",\n      \"ecr:DescribeImages\",\n
      \     \"ecr:BatchGetImage\",\n      \"route53:GetHostedZone\",\n      \"route53:List*\",\n
      \     \"route53:ListResourceRecordSets\",\n      \"route53:ChangeResourceRecordSets\"\n
      \   ],\n    \"Resource\": [\"*\"] \n  },\n  {\n    \"Effect\": \"Allow\",\n
      \   \"Action\": [\n      \"logs:CreateLogGroup\",\n      \"logs:CreateLogStream\",\n
      \     \"logs:PutLogEvents\",\n      \"logs:DescribeLogStreams\"\n    ],\n    \"Resource\":
      \"arn:aws:logs:*:*:*\"\n  }\n]\n"
  api:
    loadBalancer:
      sslCertificate:
      class: Network
      type: Internal
  authentication:
    aws:
      backendMode: CRD,MountedFile
      cpuLimit: 800m
  authorization:
    rbac: {}
  certManager:
    enabled: true
  channel: stable
  cloudLabels:
  cloudProvider: aws
  clusterAutoscaler:
    enabled: true
    skipNodesWithLocalStorage: false
    skipNodesWithSystemPods: true
    cpuRequest: "100m"
    memoryRequest: "300Mi"
  configBase: s3://
  containerRuntime: containerd
  dnsZone: 
  etcdClusters:
  - etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-us-west-2a
      name: a
    - encryptedVolume: true
      instanceGroup: master-us-west-2b
      name: b
    - encryptedVolume: true
      instanceGroup: master-us-west-2c
      name: c
    manager:
      env:
        - name: ETCD_CIPHER_SUITES
          value: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
    name: main
  - etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-us-west-2a
      name: a
    - encryptedVolume: true
      instanceGroup: master-us-west-2b
      name: b
    - encryptedVolume: true
      instanceGroup: master-us-west-2c
      name: c
    manager:
      env:
        - name: ETCD_CIPHER_SUITES
          value: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeAPIServer:
    apiAudiences:
    - api
    - istio-ca
    auditLogMaxAge: 30
    auditLogMaxBackups: 10
    auditLogMaxSize: 100
    auditLogPath: /var/log/kube-apiserver-audit.log
    disableBasicAuth: false
    enableAdmissionPlugins:
    - NamespaceLifecycle
    - LimitRanger
    - ServiceAccount
    - PersistentVolumeLabel
    - DefaultStorageClass
    - DefaultTolerationSeconds
    - MutatingAdmissionWebhook
    - ValidatingAdmissionWebhook
    - NodeRestriction
    - PodPreset
    - Priority
    - ResourceQuota
    - StorageObjectInUseProtection
    enableAggregatorRouting: true
    enableProfiling: false
    featureGates:
      EphemeralContainers: "true"
      TTLAfterFinished: "true"
    runtimeConfig:
      batch/v2alpha1: "true"
      settings.k8s.io/v1alpha1: "true"
    serviceAccountIssuer: kubernetes.default.svc
    tlsCipherSuites:
    - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
    - TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
    - TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
    - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
    - TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
    - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
    - TLS_RSA_WITH_AES_256_GCM_SHA384
    - TLS_RSA_WITH_AES_128_GCM_SHA256
  kubeControllerManager:
    authenticationKubeconfig: /var/lib/kube-controller-manager/kubeconfig
    authorizationKubeconfig: /var/lib/kube-controller-manager/kubeconfig
    featureGates:
      TTLAfterFinished: "true"
  kubeScheduler:
    authenticationKubeconfig: /var/lib/kube-scheduler/kubeconfig
    authorizationKubeconfig: /var/lib/kube-scheduler/kubeconfig
    enableProfiling: false
    featureGates:
      EphemeralContainers: "true"
  kubeDNS:
    provider: CoreDNS
    nodeLocalDNS:
      enabled: true
      memoryRequest: 5Mi
      cpuRequest: 25m
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook
    kubeReserved:
        cpu: 750m
        memory: "2Gi"
        ephemeral-storage: "1Gi"
    kubeReservedCgroup: "/kube-reserved"
    kubeletCgroups: "/kube-reserved"
    runtimeCgroups: "/kube-reserved"
    enforceNodeAllocatable: "pods,kube-reserved"
    featureGates:
      EphemeralContainers: "true"
      TTLAfterFinished: "true"
    tlsCipherSuites:
    - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
    - TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
    - TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
    - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
    - TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
    - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
    - TLS_RSA_WITH_AES_256_GCM_SHA384
    - TLS_RSA_WITH_AES_128_GCM_SHA256
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubeProxy:
    enabled: false
  kubernetesVersion: 1.20.6
  masterInternalName: api.internal.k8s.ops.us-west-2.aws.domain
  masterPublicName: api.k8s.ops.us-west-2.aws.domain
  metricsServer:
    enabled: true
    insecure: false
  networkCIDR: 172.16.0.0/16
  networkID: vpc-
  networking:
    kuberouter: {}
  nodeTerminationHandler:
    enabled: true
  nonMasqueradeCIDR: 100.64.0.0/10
  rollingUpdate:
    maxSurge: 2
  sshAccess:
  sshKeyName:
  subnets:
  - cidr: 172.16.3.0/24
    id: subnet-
    name: us-west-2a
    type: Private
    zone: us-west-2a
  - cidr: 172.16.4.0/24
    id: subnet-
    name: us-west-2b
    type: Private
    zone: us-west-2b
  - cidr: 172.16.5.0/24
    id: subnet-
    name: us-west-2c
    type: Private
    zone: us-west-2c
  - cidr: 172.16.0.0/24
    id: subnet-
    name: utility-us-west-2a
    type: Utility
    zone: us-west-2a
  - cidr: 172.16.1.0/24
    id: subnet-
    name: utility-us-west-2b
    type: Utility
    zone: us-west-2b
  - cidr: 172.16.2.0/24
    id: subnet-
    name: utility-us-west-2c
    type: Utility
    zone: us-west-2c
  topology:
    dns:
      type: Public
    masters: private
    nodes: private

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-02-01T15:20:03Z"
  generation: 5
  labels:
    kops.k8s.io/cluster: k8s.ops.us-west-2.aws.domain
  name: master-us-west-2a
spec:
  additionalSecurityGroups:
  associatePublicIp: false
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210415
  machineType: m5.xlarge
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2a
  role: Master
  rootVolumeEncryption: true
  rootVolumeType: gp3
  subnets:
  - us-west-2a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-02-01T15:20:03Z"
  generation: 5
  labels:
    kops.k8s.io/cluster: k8s.ops.us-west-2.aws.domain
  name: master-us-west-2b
spec:
  additionalSecurityGroups:
  associatePublicIp: false
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210415
  machineType: m5.xlarge
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2b
  role: Master
  rootVolumeEncryption: true
  rootVolumeType: gp3
  subnets:
  - us-west-2b

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-02-01T15:20:03Z"
  generation: 5
  labels:
    kops.k8s.io/cluster: k8s.ops.us-west-2.aws.domain
  name: master-us-west-2c
spec:
  additionalSecurityGroups:
  associatePublicIp: false
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210415
  machineType: m5.xlarge
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2c
  role: Master
  rootVolumeEncryption: true
  rootVolumeType: gp3
  subnets:
  - us-west-2c

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2020-01-31T13:54:14Z"
  generation: 2
  labels:
    kops.k8s.io/cluster: k8s.ops.us-west-2.aws.domain
  name: nodes-us-west-2a
spec:
  additionalSecurityGroups:
  cloudLabels:
    k8s.io/cluster-autoscaler/enabled: "true"
    k8s.io/cluster-autoscaler/k8s.ops.us-west-2.aws.domain: "true"
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210415
  machineType: m5n.2xlarge
  maxSize: 8
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-us-west-2a
  role: Node
  rootVolumeEncryption: true
  rootVolumeSize: 200
  rootVolumeType: gp3
  subnets:
  - us-west-2a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2020-01-31T13:55:23Z"
  generation: 2
  labels:
    kops.k8s.io/cluster: k8s.ops.us-west-2.aws.domain
  name: nodes-us-west-2b
spec:
  additionalSecurityGroups:
  cloudLabels:
    k8s.io/cluster-autoscaler/enabled: "true"
    k8s.io/cluster-autoscaler/k8s.ops.us-west-2.aws.domain: "true"
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210415
  machineType: m5n.2xlarge
  maxSize: 8
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-us-west-2b
  role: Node
  rootVolumeEncryption: true
  rootVolumeSize: 200
  rootVolumeType: gp3
  subnets:
  - us-west-2b

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2020-01-31T13:55:57Z"
  generation: 2
  labels:
    kops.k8s.io/cluster: k8s.ops.us-west-2.aws.domain
  name: nodes-us-west-2c
spec:
  additionalSecurityGroups:
  cloudLabels:
    k8s.io/cluster-autoscaler/enabled: "true"
    k8s.io/cluster-autoscaler/k8s.ops.us-west-2.aws.domain: "true"
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210415
  machineType: m5n.2xlarge
  maxSize: 8
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-us-west-2c
  role: Node
  rootVolumeEncryption: true
  rootVolumeSize: 200
  rootVolumeType: gp3
  subnets:
  - us-west-2c

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: k8s.ops.us-west-2.aws.domain
  name: nodes-m5d-us-west-2a
spec:
  additionalSecurityGroups:
  cloudLabels:
    k8s.io/cluster-autoscaler/enabled: "true"
    k8s.io/cluster-autoscaler/k8s.ops.us-west-2.aws.domain: "true"
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210415
  machineType: m5dn.2xlarge
  maxSize: 8
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-m5d-us-west-2a
  taints:
    - dedicated=concourseworker:NoSchedule
  role: Node
  rootVolumeEncryption: true
  rootVolumeSize: 200
  rootVolumeType: gp3
  subnets:
  - us-west-2a
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

**9. Anything else do we need to know?**

### PR DESCRIPTION
**Pull request by olemarkus**
Fixes #11607
It's only partial though due to https://github.com/kubernetes/kubernetes/issues/101189

If this looks good, I'll add some release notes as well

### CODE REVIEW

File: nodeup/pkg/model/kubelet.go [hunk 1]
```diff
@@ -156,6 +158,68 @@ func (b *KubeletBuilder) Build(c *fi.ModelBuilderContext) error {
...
+		{
+			cgroup := kubeletConfig.RuntimeCgroups
+			if cgroup != "" {
+				c.EnsureTask(b.buildCgroupService(cgroup))
+			}
+			manifest := &systemd.Manifest{}
+			manifest.Set("Service", "Slice", strings.Trim(cgroup, "/")+".slice")
+			manifestString := manifest.Render()
+			c.AddTask(&nodetasks.File{
+				Path:           path.Join(systemdPath, "kubelet.service.d", "10-cgroup.conf"),
```
Comment (johngmyers) [inline nodeup/pkg/model/kubelet.go:189]: Why build drop-ins to unit files we already construct ourselves? Can't we just drop the slice configuration directly into containerd.service and kubelet.service?

**Comment by johngmyers** [comment]
I've found that for k8s 1.20 that I need to create a oneshot unit to manually create the `cpuset` and `hugetlb` cgroups because the Ubuntu systemd doesn't and k8s requires they be created.

File: nodeup/pkg/model/kubelet.go (hunk 1 above)
Comment (olemarkus) [reply nodeup/pkg/model/kubelet.go:182]: Good point. I'll see if I can fix that.

**Comment by olemarkus** [comment]
Both cpuset and hugetlb is created with the current systemd version at least.

**Comment by johngmyers** [comment]
They weren't created with the systemd in Ubuntu 20.04, which is the latest LTS and the version we publish in the channels.

Perhaps this is a followup PR, but I think we'll want to include the workaround oneshot unit, as it doesn't do any harm on newer systemd.

**Pull request merged**